File: taxitips-backend/core/sources/vasttrafik.py
"""
Västtrafik (Göteborg) trafiksituationer -> den delade normaliserade alert-formen.

Port av worker/src/vasttrafik.js. Göteborg (~1,7 miljoner i regionen) är
Sveriges andra stad och ett hårt hål i produkten: Trafiklab har statisk data
för Västtrafik men ingen realtid, så operatörens egen portal är enda vägen
till levande störningar där.

Till skillnad från SL kräver detta autentisering: OAuth2 client-credentials
mot ext-api.vasttrafik.se, och applikationen måste vara PRENUMERERAD på varje
API separat i utvecklarportalen. Ett giltigt nyckelpar som saknar
prenumeration mintar ändå en token -- den ger bara 403 på varje anrop, med
`scope: default`. Det är tecknet på en saknad prenumeration, inte en trasig
hemlighet -- se authed_get.
"""

from __future__ import annotations

import base64
import math
import time
from datetime import datetime, timedelta, timezone as dt_timezone

import requests

TOKEN_URL = "https://ext-api.vasttrafik.se/token"
SITUATIONS_URL = "https://ext-api.vasttrafik.se/ts/v1/traffic-situations"
# includeGeometry=true krävs: utan den kommer varje hållplatsområde tillbaka
# med geometry: null (verifierat mot live-registret -- 0 av 11 181 bär en
# koordinat utan flaggan, alla 11 181 gör det med den).
STOPAREAS_URL = "https://ext-api.vasttrafik.se/geo/v3/StopAreas?includeGeometry=true"


def sweref99_to_wgs84(easting: float, northing: float) -> tuple[float, float]:
    """
    Västtrafik publicerar koordinater i SWEREF99TM (EPSG:3006), Sveriges
    nationella projicerade rikssystem -- t.ex. Brunnsparken är
    POINT (319346 6400119), meter från en falsk origo, INTE grader.

    Portad som den handskrivna inversa transversella Mercator-formeln,
    rad för rad -- inte utbytt mot pyproj. Formeln är redan bevisad mot
    268/268 verkliga referenspunkter i Node; ett nytt bibliotek här skulle
    byta en riskfri 1:1-port mot ett overifierat beroende, för exakt den
    sortens känslig-att-få-fel matematik där en koordinatbugg i Göteborg
    tyst skulle placera en förare på fel gathörn. Verifierad mot
    Brunnsparken i core/test_vasttrafik.py.
    """
    axis = 6378137.0  # GRS80 storaxel
    flat = 1 / 298.257222101  # GRS80 flattening
    k0, fe, lambda_zero = 0.9996, 500000.0, math.radians(15.0)

    e2 = flat * (2 - flat)
    n = flat / (2 - flat)
    a_roof = (axis / (1 + n)) * (1 + (n * n) / 4 + (n**4) / 64)

    delta1 = n / 2 - (2 * n * n) / 3 + (37 * n**3) / 96 - (n**4) / 360
    delta2 = (n * n) / 48 + (n**3) / 15 - (437 * n**4) / 1440
    delta3 = (17 * n**3) / 480 - (37 * n**4) / 840
    delta4 = (4397 * n**4) / 161280

    a_star = e2
    b_star = (5 * e2**2 - e2**3) / 6
    c_star = (104 * e2**3 - 45 * e2**4) / 120
    d_star = (1237 * e2**4) / 1260

    xi = northing / (k0 * a_roof)
    eta = (easting - fe) / (k0 * a_roof)

    xi_prim = (
        xi
        - delta1 * math.sin(2 * xi) * math.cosh(2 * eta)
        - delta2 * math.sin(4 * xi) * math.cosh(4 * eta)
        - delta3 * math.sin(6 * xi) * math.cosh(6 * eta)
        - delta4 * math.sin(8 * xi) * math.cosh(8 * eta)
    )
    eta_prim = (
        eta
        - delta1 * math.cos(2 * xi) * math.sinh(2 * eta)
        - delta2 * math.cos(4 * xi) * math.sinh(4 * eta)
        - delta3 * math.cos(6 * xi) * math.sinh(6 * eta)
        - delta4 * math.cos(8 * xi) * math.sinh(8 * eta)
    )

    phi_star = math.asin(math.sin(xi_prim) / math.cosh(eta_prim))
    delta_lambda = math.atan(math.sinh(eta_prim) / math.cos(xi_prim))
    s = math.sin(phi_star)
    phi = phi_star + s * math.cos(phi_star) * (
        a_star + b_star * s**2 + c_star * s**4 + d_star * s**6
    )

    return math.degrees(phi), math.degrees(lambda_zero + delta_lambda)


# Token-cachning speglar Node-workerns fcmPush.js-mönster: tokens gäller
# ~24h, så att minta en per anrop vore rent slöseri. OBS: den här cachen
# hjälper bara INOM en process -- varje `manage.py poll_vasttrafik`-körning
# är en ny process och mintar sin egen token. Accepterad begränsning,
# matchar kodbasens "ingen Celery, medvetet minimal"-hållning; blir
# värdefull den dag ett långlivat Django-processer finns.
_cached_token: str | None = None
_cached_token_expires_at: float = 0


def get_access_token(client_id: str, client_secret: str) -> str:
    global _cached_token, _cached_token_expires_at
    if _cached_token and time.time() < _cached_token_expires_at - 60:
        return _cached_token

    basic = base64.b64encode(f"{client_id}:{client_secret}".encode()).decode()
    res = requests.post(
        TOKEN_URL,
        headers={"Authorization": f"Basic {basic}", "Content-Type": "application/x-www-form-urlencoded"},
        data="grant_type=client_credentials",
        timeout=30,
    )
    if not res.ok:
        raise RuntimeError(f"vasttrafik token {res.status_code}: {res.text[:160]}")
    body = res.json()
    if not body.get("access_token"):
        raise RuntimeError("vasttrafik token: no access_token in response")
    _cached_token = body["access_token"]
    _cached_token_expires_at = time.time() + float(body.get("expires_in") or 3600)
    return _cached_token


def authed_get(url: str, client_id: str, client_secret: str) -> dict:
    token = get_access_token(client_id, client_secret)
    res = requests.get(
        url, headers={"Authorization": f"Bearer {token}", "Accept-Encoding": "gzip"}, timeout=30
    )
    if res.status_code == 403:
        # Skiljer "inte prenumererad" från "fel autentisering": en token
        # mintades redan ovan, så nyckelparet är giltigt -- applikationen är
        # bara inte prenumererad på det här API:t i utvecklarportalen. Att
        # säga det rakt ut sparar en omtestning av hemligheten.
        from urllib.parse import urlparse
        path = urlparse(url).path
        raise RuntimeError(
            f"vasttrafik 403 on {path} -- token minted OK, so the application "
            f"is likely not subscribed to this API. Subscribe it in "
            f"developer.vasttrafik.se (Störning v1 / Geografi v3) and retry."
        )
    if not res.ok:
        raise RuntimeError(f"vasttrafik {res.status_code}: {res.text[:160]}")
    return res.json()


# Västtrafik anger start/sluttid för själva situationen (startTime/endTime),
# en genuin förbättring över SL:s publiceringsfönster -- den beskriver
# störningen, inte meddelandet. Samma agerbarhetsregel gäller ändå: en
# månadslång planerad avstängning är ett faktum om kartan, inte ett
# taxitillfälle.
MAX_AGE = timedelta(hours=24)
MAX_WINDOW = timedelta(days=7)


def _parse(value: str | None) -> datetime | None:
    if not value:
        return None
    try:
        return datetime.fromisoformat(value)
    except ValueError:
        return None


def is_actionable(situation: dict, now: datetime | None = None) -> bool:
    now = now or datetime.now(dt_timezone.utc)
    start = _parse(situation.get("startTime"))
    end = _parse(situation.get("endTime"))
    if start is None or end is None:
        return False
    # Måste ha PÅBÖRJATS. Västtrafik publicerar planerat spårarbete långt i
    # förväg -- mätt: 17 av 87 situationer, och 10 av de 19 som klarade
    # ålder/varaktighet-testet ensamt, hade inte börjat än (nattarbete
    # schemalagt upp till tre veckor ut).
    if start > now:
        return False
    if now - start > MAX_AGE:
        return False
    if end - start > MAX_WINDOW:
        return False
    return end > now


def normalize_severity(severity: str | None) -> str | None:
    """
    Västtrafiks `severity` är operatörens egen bedömning (dokumenterade
    värden: normal/high/veryHigh). Precis som SL:s importance_level bärs
    den vidare som bevis, mappas aldrig in i demand_score.
    """
    v = str(severity or "").lower()
    if v in ("veryhigh", "very_high"):
        return "veryHigh"
    if v == "high":
        return "high"
    if v == "normal":
        return "normal"
    if v == "slight":
        return "slight"
    return None


def mode_hint_from(lines: list[dict] | None) -> str | None:
    modes = {
        str(l.get("defaultTransportModeCode") or l.get("transportMode") or "").lower()
        for l in (lines or [])
    }
    modes.discard("")
    if not modes:
        return None
    if "tram" in modes:
        return "tram"
    if "train" in modes or "rail" in modes:
        return "train"
    if "bus" in modes:
        return "bus"
    if "ferry" in modes or "boat" in modes:
        return "boat"
    return None


def normalize_situation(situation: dict, stop_area_index: dict | None) -> dict | None:
    stop_points = situation.get("affectedStopPoints") or []
    # Tåg/expressbuss-situationer bär INGA affectedLines -- linjen sitter
    # under affectedJourneys[].line istället (verifierat: varje "Öresundståg
    # NNNN är inställt"-situation har affectedLines: [] och en ifylld
    # affectedJourneys). Missar man det faller inställda tåg -- den enskilt
    # mest värdefulla tipstypen -- till mode "unknown" och därmed
    # severity_tier "ignore".
    journeys = situation.get("affectedJourneys") or []
    lines = situation.get("affectedLines") or [j.get("line") for j in journeys if j.get("line")]

    areas = list(dict.fromkeys(
        [s.get("stopAreaName") or s.get("name") for s in stop_points if s.get("stopAreaName") or s.get("name")]
        + [s.get("municipalityName") for s in stop_points if s.get("municipalityName")]
    ))
    routes = list(dict.fromkeys(
        str(l.get("designation") or l.get("name")) for l in lines if l.get("designation") or l.get("name")
    ))
    stops = list(dict.fromkeys(
        str(s.get("stopAreaGid") or s.get("gid")) for s in stop_points if s.get("stopAreaGid") or s.get("gid")
    ))

    # Hållplatspunkterna i situations-flödet bär INGA koordinater -- bara en
    # stopAreaGid, som slås upp mot geo-registret (mätt: 268/268 av
    # refererade gid:n löser). Registrets egna koordinater är SWEREF99TM och
    # projiceras till WGS84 när indexet byggs.
    lat = None
    lon = None
    for sp in stop_points:
        hit = (stop_area_index or {}).get(str(sp.get("stopAreaGid"))) or (stop_area_index or {}).get(str(sp.get("gid")))
        if hit:
            lat, lon = hit["lat"], hit["lon"]
            break

    start = _parse(situation.get("startTime"))
    end = _parse(situation.get("endTime"))

    # affectedJourneys ger, när den finns (~11% av situationer, mätt live),
    # den drabbade turens egna schemalagda avgångstid och riktning -- t.ex.
    # "830, Munkedal → Uddevalla". Det är INTE "nästa avgång" (det kräver en
    # hel tidtabell att jämföra mot, som varken den här källan eller
    # Trafiklab exponerar -- se core/sources/trafikverket_rail.py:s
    # next_departure_minutes, den enda källan med det), bara vilken tur och
    # åt vilket håll det gällde.
    first_journey = journeys[0] if journeys else None
    journey_line = (first_journey or {}).get("line") or {}
    route_label = str(journey_line.get("designation") or "") or (routes[0] if routes else "")
    direction = None
    if journey_line.get("directions"):
        direction = journey_line["directions"][0].get("name")
    journey_departure_at = _parse((first_journey or {}).get("departureDateTime"))

    return {
        "id": f"vt:{situation.get('situationNumber')}",
        "header": situation.get("title") or "Störning",
        "description": situation.get("description") or "",
        "cause": None,
        "effect": None,
        "areas": areas, "routes": routes, "stops": stops,
        "route_label": route_label or None,
        "direction": direction,
        "journey_departure_at": journey_departure_at,
        "url": None,
        "active_from": start or datetime.now(dt_timezone.utc),
        "active_to": end,
        "source": "vt",
        "region": "vt",
        "lat": lat, "lon": lon,
        "mode_hint": mode_hint_from(lines),
        "vt": {"severity": normalize_severity(situation.get("severity"))},
    }


def build_stop_area_index(rows: list[dict] | None) -> dict[str, dict]:
    """gid -> {lat, lon, name} för Göteborgs hållplatsregister."""
    return {str(r["gid"]): r for r in (rows or [])}


def fetch_vasttrafik_stop_areas(client_id: str, client_secret: str) -> list[dict]:
    body = authed_get(STOPAREAS_URL, client_id, client_secret)
    areas = body.get("stopAreas") if isinstance(body, dict) else (body if isinstance(body, list) else [])

    seen: set[str] = set()
    rows = []
    for a in areas or []:
        gid = a.get("gid")
        if not gid or str(gid) in seen:
            continue
        geometry = a.get("geometry") or {}
        if geometry.get("eastingCoordinate") is None or geometry.get("northingCoordinate") is None:
            continue
        seen.add(str(gid))
        lat, lon = sweref99_to_wgs84(
            float(geometry["eastingCoordinate"]), float(geometry["northingCoordinate"])
        )
        rows.append({"gid": str(gid), "operator": "vt", "name": a.get("name"), "lat": lat, "lon": lon})
    return rows


def fetch_vasttrafik_situations(
    client_id: str | None, client_secret: str | None,
    stop_area_index: dict | None = None, now: datetime | None = None,
) -> dict:
    if not client_id or not client_secret:
        return {"alerts": [], "source": "vt", "skipped": "no credentials"}

    now = now or datetime.now(dt_timezone.utc)
    body = authed_get(SITUATIONS_URL, client_id, client_secret)
    all_situations = body if isinstance(body, list) else (body.get("results") or [])

    alerts = []
    for situation in all_situations:
        if not is_actionable(situation, now):
            continue
        alert = normalize_situation(situation, stop_area_index)
        if alert:
            alerts.append(alert)

    return {"alerts": alerts, "source": "vt", "received": len(all_situations), "actionable": len(alerts)}

File: taxitips-api/worker/src/vasttrafik.js
/**
 * Västtrafik (Göteborg) traffic situations -> the shared normalized alert shape.
 *
 * Göteborg is Sweden's second city (~1.7M in the region) and a hard gap in the
 * product: Trafiklab's own coverage tables list Västtrafik with static data but
 * NO realtime, so the operator's own portal is the only route to live
 * disruptions there.
 *
 * Unlike SL this needs credentials: OAuth2 client-credentials against
 * ext-api.vasttrafik.se, and the application must be *subscribed* to each API
 * separately in the developer portal. A key pair that is valid but unsubscribed
 * still mints a token -- it just returns 403 `900908 Resource forbidden` on
 * every call, with `scope: default` and no application name in the JWT. That is
 * the signature of a missing subscription, not a bad secret; see
 * fetchVasttrafikSituations' error handling.
 *
 * Credentials come from VASTTRAFIK_CLIENT_ID / VASTTRAFIK_CLIENT_SECRET.
 */

const TOKEN_URL = "https://ext-api.vasttrafik.se/token";
const SITUATIONS_URL = "https://ext-api.vasttrafik.se/ts/v1/traffic-situations";
// `includeGeometry=true` is required: without it every stop area comes back
// with `geometry: null` (verified against the live register -- 0 of 11,181
// carry a coordinate by default, all 11,181 do with the flag).
const STOPAREAS_URL =
  "https://ext-api.vasttrafik.se/geo/v3/StopAreas?includeGeometry=true";

/**
 * Västtrafik publishes coordinates in SWEREF99TM (EPSG:3006), Sweden's
 * national projected grid -- e.g. Brunnsparken is
 * `POINT (319346 6400119)`, metres from a false origin, NOT degrees.
 * Everything downstream (get_smart_alerts' Haversine, flutter_map, the
 * driver's distance) assumes WGS84 degrees, so these must be projected or a
 * Göteborg tip lands in the Gulf of Guinea.
 *
 * Implemented inline rather than pulling in proj4: this is one well-defined
 * transverse-Mercator inverse with fixed parameters (GRS80, central meridian
 * 15°E, scale 0.9996, false easting 500 km), and the formula is stable.
 * Accuracy is sub-metre over Sweden, far tighter than the ~100 m precision
 * that matters for "which street corner".
 */
function sweref99ToWgs84(easting, northing) {
  const axis = 6378137.0; // GRS80 semi-major
  const flat = 1 / 298.257222101; // GRS80 flattening
  const k0 = 0.9996, fe = 500000.0, lambdaZero = (15.0 * Math.PI) / 180;

  const e2 = flat * (2 - flat);
  const n = flat / (2 - flat);
  const aRoof = (axis / (1 + n)) * (1 + (n * n) / 4 + (n * n * n * n) / 64);

  const delta1 = n / 2 - (2 * n * n) / 3 + (37 * n ** 3) / 96 - (n ** 4) / 360;
  const delta2 = (n * n) / 48 + (n ** 3) / 15 - (437 * n ** 4) / 1440;
  const delta3 = (17 * n ** 3) / 480 - (37 * n ** 4) / 840;
  const delta4 = (4397 * n ** 4) / 161280;

  const Astar = e2, Bstar = (5 * e2 ** 2 - e2 ** 3) / 6;
  const Cstar = (104 * e2 ** 3 - 45 * e2 ** 4) / 120;
  const Dstar = (1237 * e2 ** 4) / 1260;

  const xi = (northing - 0) / (k0 * aRoof);
  const eta = (easting - fe) / (k0 * aRoof);

  const xiPrim =
    xi -
    delta1 * Math.sin(2 * xi) * Math.cosh(2 * eta) -
    delta2 * Math.sin(4 * xi) * Math.cosh(4 * eta) -
    delta3 * Math.sin(6 * xi) * Math.cosh(6 * eta) -
    delta4 * Math.sin(8 * xi) * Math.cosh(8 * eta);
  const etaPrim =
    eta -
    delta1 * Math.cos(2 * xi) * Math.sinh(2 * eta) -
    delta2 * Math.cos(4 * xi) * Math.sinh(4 * eta) -
    delta3 * Math.cos(6 * xi) * Math.sinh(6 * eta) -
    delta4 * Math.cos(8 * xi) * Math.sinh(8 * eta);

  const phiStar = Math.asin(Math.sin(xiPrim) / Math.cosh(etaPrim));
  const deltaLambda = Math.atan(Math.sinh(etaPrim) / Math.cos(xiPrim));
  const s = Math.sin(phiStar);
  const phi =
    phiStar +
    s * Math.cos(phiStar) *
      (Astar +
        Bstar * s ** 2 +
        Cstar * s ** 4 +
        Dstar * s ** 6);

  return {
    lat: (phi * 180) / Math.PI,
    lon: ((lambdaZero + deltaLambda) * 180) / Math.PI,
  };
}

/**
 * Token caching mirrors fcmPush.js's getAccessToken: tokens are valid for
 * ~24h, so minting one per 60s poll would be pure waste (and rate-limit
 * bait). Refresh a minute early to avoid using one that expires mid-request.
 */
let cachedToken = null;
let cachedTokenExpiresAt = 0;

async function getAccessToken(clientId, clientSecret) {
  if (cachedToken && Date.now() < cachedTokenExpiresAt - 60_000) return cachedToken;

  const basic = Buffer.from(`${clientId}:${clientSecret}`).toString("base64");
  const res = await fetch(TOKEN_URL, {
    method: "POST",
    headers: {
      Authorization: `Basic ${basic}`,
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: "grant_type=client_credentials",
  });
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(`vasttrafik token ${res.status}: ${body.slice(0, 160)}`);
  }
  const json = await res.json();
  if (!json.access_token) throw new Error("vasttrafik token: no access_token in response");
  cachedToken = json.access_token;
  cachedTokenExpiresAt = Date.now() + Number(json.expires_in || 3600) * 1000;
  return cachedToken;
}

async function authedGet(url, clientId, clientSecret) {
  const token = await getAccessToken(clientId, clientSecret);
  const res = await fetch(url, {
    headers: { Authorization: `Bearer ${token}`, "Accept-Encoding": "gzip" },
  });
  if (res.status === 403) {
    // Distinguish "not subscribed" from "bad credentials": a token was minted
    // successfully above, so the key pair is valid -- the application just
    // isn't subscribed to this API in the developer portal. Saying so plainly
    // saves re-testing the secret.
    throw new Error(
      `vasttrafik 403 on ${new URL(url).pathname} -- token minted OK, so the ` +
        `application is likely not subscribed to this API. Subscribe it in ` +
        `developer.vasttrafik.se (Störning v1 / Geografi v3) and retry.`
    );
  }
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(`vasttrafik ${res.status}: ${body.slice(0, 160)}`);
  }
  return res.json();
}

/**
 * Västtrafik states start/end times for the situation itself (startTime /
 * endTime), which is a genuine improvement over SL's publish window -- it
 * describes the disruption rather than the message. The same actionability
 * rule still applies though: a months-long planned closure is a fact about
 * the map, not a taxi opportunity.
 */
const MAX_AGE_MS = 24 * 60 * 60 * 1000;
const MAX_WINDOW_MS = 7 * 24 * 60 * 60 * 1000;

function isActionable(situation, now = Date.now()) {
  const from = Date.parse(situation?.startTime || "");
  const to = Date.parse(situation?.endTime || "");
  if (!Number.isFinite(from)) return false;
  if (!Number.isFinite(to)) return false;
  // Must have STARTED. Västtrafik publishes planned engineering work well in
  // advance -- measured: 17 of 87 situations, and 10 of the 19 that passed
  // the age/duration test alone, had not begun yet (night-time track work
  // scheduled up to three weeks out). A tram line that goes down next
  // Saturday is not a taxi opportunity tonight, and showing it as one is the
  // same false-tip failure the SL publish-window filter guards against.
  if (from > now) return false;
  if (now - from > MAX_AGE_MS) return false;
  if (to - from > MAX_WINDOW_MS) return false;
  return to > now;
}

/**
 * Västtrafik's `severity` is the operator's own judgement (documented values
 * include normal/high/veryHigh). Like SL's importance_level it is carried
 * through as evidence, never mapped into demand_score -- it ranks messages
 * for Västtrafik's display surfaces, it does not measure taxi demand.
 */
function normalizeSeverity(severity) {
  // Measured on the live feed: only `normal` (14) and `slight` (73) appear,
  // but the model documents higher levels too. Kept as reported.
  const v = String(severity || "").toLowerCase();
  if (v === "veryhigh" || v === "very_high") return "veryHigh";
  if (v === "high") return "high";
  if (v === "normal") return "normal";
  if (v === "slight") return "slight";
  return null;
}

function modeHintFrom(lines) {
  const modes = new Set(
    (Array.isArray(lines) ? lines : [])
      .map((l) =>
        String(l?.defaultTransportModeCode || l?.transportMode || "").toLowerCase()
      )
      .filter(Boolean)
  );
  if (!modes.size) return null;
  if (modes.has("tram")) return "tram";
  if (modes.has("train") || modes.has("rail")) return "train";
  if (modes.has("bus")) return "bus";
  if (modes.has("ferry") || modes.has("boat")) return "boat";
  return null;
}

function normalizeSituation(situation, stopAreaIndex) {
  const stopPoints = Array.isArray(situation.affectedStopPoints)
    ? situation.affectedStopPoints
    : [];
  // Train/coach situations carry no `affectedLines` at all -- the line sits
  // under `affectedJourneys[].line` instead (verified: every "Öresundståg
  // NNNN är inställt" situation has affectedLines: [] and a populated
  // affectedJourneys). Missing this drops cancelled trains, which are the
  // single most valuable tip type there is, to mode "unknown" and therefore
  // to severity_tier "ignore".
  const journeys = Array.isArray(situation.affectedJourneys)
    ? situation.affectedJourneys
    : [];
  const lines = (
    Array.isArray(situation.affectedLines) && situation.affectedLines.length
      ? situation.affectedLines
      : journeys.map((j) => j?.line).filter(Boolean)
  );

  const areas = [
    ...new Set(
      stopPoints
        .map((s) => s.stopAreaName || s.name)
        .filter(Boolean)
        .concat(stopPoints.map((s) => s.municipalityName).filter(Boolean))
    ),
  ];
  const routes = [
    ...new Set(lines.map((l) => l.designation || l.name).filter(Boolean).map(String)),
  ];
  const stops = [
    ...new Set(stopPoints.map((s) => String(s.stopAreaGid || s.gid)).filter(Boolean)),
  ];

  // Stop points in the situations feed carry NO coordinates -- only a
  // `stopAreaGid`, which joins the geo register (measured: 268/268 of the
  // referenced gids resolve). The register's own coordinates are SWEREF99TM
  // and are projected to WGS84 when the index is built.
  let lat = null;
  let lon = null;
  for (const sp of stopPoints) {
    const hit =
      stopAreaIndex?.get(String(sp.stopAreaGid)) ||
      stopAreaIndex?.get(String(sp.gid));
    if (hit) {
      lat = hit.lat;
      lon = hit.lon;
      break;
    }
  }

  const from = Date.parse(situation.startTime || "");
  const to = Date.parse(situation.endTime || "");

  return {
    id: `vt:${situation.situationNumber}`,
    header: situation.title || "Störning",
    description: situation.description || "",
    cause: null,
    effect: null,
    areas,
    routes,
    stops,
    url: null,
    active_from: Number.isFinite(from) ? from : Date.now(),
    active_to: Number.isFinite(to) ? to : null,
    source: "vt",
    region: "vt",
    lat,
    lon,
    modeHint: modeHintFrom(lines),
    vt: { severity: normalizeSeverity(situation.severity) },
    // Västtrafik states the disruption's own start/end, unlike SL's publish
    // window -- so active_from/active_to above are the real thing.
  };
}

/** gid -> {lat, lon, name} for the Göteborg stop-area register. */
function buildStopAreaIndex(rows) {
  return new Map((rows || []).map((r) => [String(r.gid), r]));
}

async function fetchVasttrafikStopAreas(clientId, clientSecret) {
  const json = await authedGet(STOPAREAS_URL, clientId, clientSecret);
  const areas = json?.stopAreas || (Array.isArray(json) ? json : []);
  // Deduplicated for the same reason as sl.js: an upsert batch may not touch
  // the same key twice.
  const seen = new Set();
  const rows = [];
  for (const a of areas) {
    if (a?.gid && seen.has(String(a.gid))) continue;
    if (a?.gid) seen.add(String(a.gid));
    const g = a?.geometry;
    if (!a?.gid || !g || g.eastingCoordinate == null || g.northingCoordinate == null) continue;
    const { lat, lon } = sweref99ToWgs84(
      Number(g.eastingCoordinate),
      Number(g.northingCoordinate)
    );
    rows.push({ gid: String(a.gid), name: a.name || null, lat, lon });
  }
  return rows;
}

async function fetchVasttrafikSituations({
  clientId = process.env.VASTTRAFIK_CLIENT_ID,
  clientSecret = process.env.VASTTRAFIK_CLIENT_SECRET,
  stopAreaIndex = null,
  now = Date.now(),
} = {}) {
  if (!clientId || !clientSecret) {
    return { alerts: [], source: "vt", skipped: "no credentials" };
  }
  const json = await authedGet(SITUATIONS_URL, clientId, clientSecret);
  const all = Array.isArray(json) ? json : json?.results || [];

  const alerts = [];
  for (const situation of all) {
    if (!isActionable(situation, now)) continue;
    const alert = normalizeSituation(situation, stopAreaIndex);
    if (alert) alerts.push(alert);
  }
  return { alerts, source: "vt", received: all.length, actionable: alerts.length };
}

module.exports = {
  fetchVasttrafikSituations,
  fetchVasttrafikStopAreas,
  buildStopAreaIndex,
  isActionable,
  modeHintFrom,
  normalizeSituation,
  normalizeSeverity,
};
